Add unit tests for AddressBook class #105

thenaesh opened the pull request:
Fixes #105
Ready for review

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/se-edu/addressbook-level2/121)
<!-- Reviewable:end -->

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,359 @@
+                                   new Phone("168463354", false),
+                                   new Email("email@notinvented.yet", false),
+                                   new Address("James Monroe Tomb, Richmond, Virginia", false),
+                                   new UniqueTagList(tagIndependent, tagDead));
```
When defining test data, try to follow a pattern that is easy to remember. For example, the first person's name has initials A. B. (e.g. Alice Betsy), second person's name has initials B. C. (Ben Charles) and so on. When things always follow a pattern, anything going out of that pattern (e.g. Ben suddenly appears at the start of the list) gets noticed immediately and sometimes can lead to discovering bugs.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,359 @@
+ */
+public class AddressBookTest {
+	
+	private Tag tagDead;
```
Wrong indentation. Should be 4 spaces.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,359 @@
+	}
+	
+	/**
+	 * checks if the underlying container behind an iterable is empty
```
`Checks`

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,359 @@
+import seedu.addressbook.data.tag.UniqueTagList.TagNotFoundException;
+
+/**
+ * we check, in particular, for the two guarantees specified by the AddressBook class
```
`We`

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,359 @@
+	private Person personBarack; // has tagDemocrat
+	private Person personGeorge; // has tagRepublican
+	private Person personBill;   // has tagDemocrat
+	private Person personMonroe; // has tagIndependent, tagDead
```
OSS projects are supposed to last long and cross culture boundaries. So it is best not to pick names/values that may not be known to all or may be known as well after some time.

damithc replied:
In other words, code should not give subtle hints as to the period in which it was written or the person who wrote it.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,359 @@
+/**
+ * we check, in particular, for the two guarantees specified by the AddressBook class
+ *  - Every tag found in every person will also be found in the tag list.
+ *  - The tags in each person point to tag objects in the master list. (== equality)
```
By convention this class tests the AddressBook class. So there is no need to explain what functionality it tests. It should test _all_ functionality of the AddressBook class. Tests can be organized based on public methods of the class such that for each public method there is one or more test methods.

thenaesh commented:
@damithc I've made changes taking into account all the comments mentioned. I've removed the redundant comment at the start of the test class, fixed the indentation and some other formatting. I've also made the test cases culturally-neutral and also made them follow the pattern suggested.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,372 @@
+    }
+
+    @Test
+    public void addTag_addDuplicate_throwsDuplicateTagException() throws Exception {
```
`addTag_tagAlreadyInList_throwsDuplicateTagException`

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,372 @@
+        boolean foundPerson = addressBook.containsPerson(personCharlieDouglas);
+
+        assertFalse(foundPerson);
+    }
```
Can be shorter
```java
@Test
public void containsPerson() throws Exception {
    AddressBook addressBook = createDefaultAddressBook();
    assertTrue(addressBook.containsPerson(personAliceBetsy));
    assertFalse(addressBook.containsPerson(personCharlieDouglas));
}
```

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,372 @@
+
+        // UniqueTagList constructor will not throw as there are no duplicate
+        // elements
+        personAliceBetsy     = new Person(new Name("Alice Betsy"),
```
`person` prefix can be omitted.

thenaesh commented:
@damithc I've moved the helper methods out into the TestUtils class and cleaned up the unnecessary comments. I've also combined, renamed and simplified tests according to the suggestions.

I've modified the tests expecting exceptions to be thrown to use JUnit's ExpectedException instead of manually catching the exceptions. As PR #120 has not been merged yet, I am not using PersonBuilder to create the Person objects at the moment.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,275 @@
+    /**
+     * Initialises an AddressBook with some default values.
+     * 
+     * Contains personAliceBetsy and personBobChaplin.
```
`person` prefix should be deleted.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,275 @@
+
+        assertTrue(addressBook.containsTag(tagMathematician));
+        assertTrue(addressBook.containsTag(tagScientist));
+        assertFalse(addressBook.containsTag(tagEconomist));
```
Can this be done in one shot (and a more definitive way)? e.g. 
`assertEquals(expectedTagList, addressBook.getAllTags())`

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,275 @@
+        }
+        
+        assertTrue(allPersonTagsValid);
+    }
```
What does this test case tests that the previous methods don't?
I think the only functionality we can test is when a person is added, if that person has a tag that is missing in the tag list, it will be added to the list. In fact, it can be tested while testing `addPerson` method.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,275 @@
+        AddressBook addressBook = createDefaultAddressBook();
+        
+        thrown.expect(DuplicatePersonException.class);
+        addressBook.addPerson(aliceBetsy);
```
Here, you can verify that if the duplicate person has a tag that is not in the list, it will not be added to the list.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,275 @@
+
+    @Test
+    public void addTag_tagNotInList_addsNormally() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
```
Looks like this line appears in every `@Test`. May be put that in the `@Before` method?

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,275 @@
+    public void containsPerson() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+
+        assertTrue(addressBook.containsPerson(aliceBetsy));
```
Loop over all Persons that is supposed to be in the list?
You should also test this method for empty list.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,275 @@
+        AddressBook addressBook = createDefaultAddressBook();
+
+        assertTrue(addressBook.containsPerson(aliceBetsy));
+        addressBook.removePerson(aliceBetsy);
```
Should at least check the size to ensure no more than the target person was removed.

damithc commented on test/java/seedu/addressbook/util/TestUtil.java:
```diff
@@ -51,6 +56,34 @@ public static AddressBook createAddressBook(Person... persons) throws DuplicateP
     }
+
+    /**
+     * Returns true iff every pair of corresponding elements two iterables are (deeply) identical.
```
The method is order-sensitive? If yes, should mention in the header comment.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,275 @@
+    private Person aliceBetsy; // has tagMathematician
+    private Person bobChaplin; // has tagMathematician
+    private Person charlieDouglas; // has tagScientist
+    private Person davidElliot; // has tagEconomist, tagPrizeWinner
```
The info in these comments are also present in another place (in the `setUp` method), and therefore this is a violation of the DRY principle. Comments can be removed.

thenaesh commented:
@damithc I've made all the changes requested, including spreading out the tag equality tests.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,289 @@
+            tagEqualityPropertyHolds = tagEqualityPropertyHolds || (tag == tagToCheck);
+        }
+        
+        return tagEqualityPropertyHolds;
```
Alternative algo (shorter):
```java
for (Tag tag : addressBook.getAllTags()) {
    if (tag == tagToCheck) { return true; }
}

return false;
```

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,289 @@
+                                    new Address("11 Arts Link", false),
+                                    new UniqueTagList(tagEconomist, tagPrizeWinner));
+        
+        defaultAddressBook = createDefaultAddressBook();
```
Use this instead?
```java
defaultAddressBook = new AddressBook(new UniquePersonList(aliceBetsy, bobChaplin),
                                     new UniqueTagList(tagMathematician, tagScientist));

damithc commented on src/seedu/addressbook/data/exception/ConstraintViolationException.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Signals that the operation should not take place as it will violate some constraint.
+ */
+public class ConstraintViolationException extends Exception {
```
Why are you introducing a new exception just for testing?

thenaesh commented:
@damithc I've made all the changes requested, except for the `ConstraintViolationException`. I'm quite confused with the implementation of `removeTag` in the `AddressBook` class. The guarantees specified in comments at the top of the file state that "every tag found in every person will also be found in the tag list", but `removeTag` will happily delete any tag as long as it can find it in the tag list, even if there is still a person with that tag in the address book.

I'm not sure if I'm misunderstanding the meaning of the guarantee, or if this behaviour is actually a bug. If it is the former, perhaps you could clarify and I can then correct the test case. If it is the latter, and if the abovementioned scenario arises, shouldn't `removeTag` throw an exception that is something other than `TagNotFoundException`? I introduced a new exception because this seemed to be the case. If I shouldn't introduce a new exception, then should I just test for `Exception`, or do something else entirely?

damithc commented:
@thenaesh that sounds like a bug. In that case, it may be ok to throw the exception.  Go ahead.

In the meantime, @YijinL any input from you?

thenaesh commented:
I've added a commit fixing that bug.

There's also another failing test `addPerson_personAlreadyInListButHasTagNotInList_tagNotAdded`, which fails because `addPerson` does a `syncTagsWithMasterList` before attempting to actually add the `Person` (which may throw). I've also added another commit fixing that bug (which is simply swapping the two lines around).

I can revert the commits if the fixes are not desired (or if I should perform them in a separate PR).

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,253 @@
+    }
+    
+    
+    // Helper Methods
```
No need

damithc commented on test/java/seedu/addressbook/util/TestUtil.java:
```diff
@@ -51,6 +56,48 @@ public static AddressBook createAddressBook(Person... persons) throws DuplicateP
+    }
+    
+    /**
+     * Returns true iff the underlying container behind an iterable is empty.
```
Do we need the `iff`? In other places we use `if` even when we mean `if and only if`

damithc commented on test/java/seedu/addressbook/util/TestUtil.java:
```diff
@@ -51,6 +56,48 @@ public static AddressBook createAddressBook(Person... persons) throws DuplicateP
+        }
+        
+        return numberOfElementsSeen;
+    }
```
I'm a bit sceptical that we have to write all these utility methods for iterators. Google a bit to see if there are standard solutions.

thenaesh replied:
It seems that there are no standard solutions that do not involve adding new dependencies (e.g. Guava).

damithc commented on src/seedu/addressbook/data/AddressBook.java:
```diff
@@ -118,7 +119,12 @@ public void removePerson(ReadOnlyPerson toRemove) throws PersonNotFoundException
+            if (person.getTags().contains(toRemove)) {
+                throw new ConstraintViolationException("The Tag to be removed is present in a Person object.");
+            }
+        }
```
Extract a `isTagUsed()` method to improve SLAP?

YijinL commented:
Yes, `removeTag` should check and remove the appropriate tag from all persons in the addressbook.

damithc commented:
> Yes, removeTag should check and remove the appropriate tag from all persons in the addressbook.
@thenaesh that sounds a better approach than the current approach. In that case we can do without the exception too.

thenaesh commented:
@damithc I've made the requested changes, except for replacing the helper methods with standard solutions (as I couldn't find any that didn't involve adding new dependencies).

damithc commented:
@thenaesh this branch is outdated. Can rebase?

thenaesh commented:
@damithc Oops, sorry forgot. Done.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,258 @@
+            return;
+        }
+        
+        fail("Expected DuplicatePersonException was not caught.");
```
This flow more intuitive?

```java
try {
     defaultAddressBook.addPerson(aliceBetsy);
     fail("Expected DuplicatePersonException was not thrown.");
} catch (DuplicatePersonException e) {
   // ignore expected exception
}

assertFalse(defaultAddressBook.containsTag(tagPrizeWinner));
       
```

damithc replied:
Resolved

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,258 @@
+        for (Tag tag : allTags) {
+            assertFalse(emptyAddressBook.containsTag(tag));
+        }
+    }
```
Why the above two methods have different structures? They are testing similar things right?

damithc replied:
Resolved

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,258 @@
+    public void removePerson_personExists_removesNormally() throws Exception {
+        int numberOfPersonsBeforeRemoval = getSize(defaultAddressBook.getAllPersons());
+        defaultAddressBook.removePerson(aliceBetsy);
+        int numberOfPersonsAfterRemoval = getSize(defaultAddressBook.getAllPersons());
```
This line can be moved further down. Don't define a variable until you need it.

damithc replied:
Resolved

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,258 @@
+        defaultAddressBook.clear();
+
+        // If the iterator does not have next element at start, then underlying
+        // container has no elements.
```
Sounds like this comment belongs to the `isEmpty()` method.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,258 @@
+    }
+    
+    /**
+     * Returns true if the given Tag object if found in the tag list of the given AddressBook.
```
`if found` -> `is found`?

damithc replied:
Resolved

damithc commented:
@thenaesh reminder

thenaesh commented:
@damithc Sorry for the delay, I've made the changes requested.

damithc commented on src/seedu/addressbook/data/AddressBook.java:
```diff
@@ -119,8 +119,21 @@ public void removePerson(ReadOnlyPerson toRemove) throws PersonNotFoundException
      * @throws TagNotFoundException if no such Tag could be found.
      */
     public void removeTag(Tag toRemove) throws TagNotFoundException {
+        removeTagFromAllPersons(toRemove);
```
The header comment can be updated to mention this behavior.

damithc commented on test/java/seedu/addressbook/data/AddressBookTest.java:
```diff
@@ -0,0 +1,265 @@
+        assertFalse(defaultAddressBook.containsTag(tagMathematician));
+        
+        int numberOfTagsAfterRemoval = getSize(defaultAddressBook.getAllTags());
+        assertTrue(numberOfTagsBeforeRemoval == numberOfTagsAfterRemoval + 1);
```
Is this more intuitive? `numberOfTagsAfterRemoval == numberOfTagsBeforeRemoval - 1`

pyokagan commented:
@damithc Sorry, but with the semester starting I don't have any more capacity for immediate reviews.

damithc commented:
> with the semester starting I don't have any more capacity for immediate reviews.

Noted @pyokagan. No worries.

thenaesh commented:
@damithc I've made the requested changes.
